lightning/src/offers/flow.rs

```diff
@@ -72,6 +72,115 @@ use {
 	crate::onion_message::dns_resolution::{DNSResolverMessage, DNSSECQuery, OMNameResolver},
 };
 
+/// Contains OfferEvents that can optionally triggered for manual
+/// handling by user based on appropriate user_config.
+pub enum OfferEvents {
+	/// Notifies that a verified [`InvoiceRequest`] has been received.
+	///
+	/// This event is triggered when a BOLT12 [`InvoiceRequest`] is received.
+	/// The `amount_source` field describes how the amount is specified — in the offer,
+	/// the invoice request, or both.
+	///
+	/// To respond, use [`OffersMessageFlow::create_invoice_builder_from_invoice_request`], providing an
+	/// amount based on the structure described in [`InvoiceRequestAmountSource`].
+	///
+	/// See the [BOLT12 spec](https://github.com/lightning/bolts/blob/master/12-offer-encoding.md#requirements-1)
+	/// for protocol-level details.
+	InvoiceRequestReceived {
+		/// The verified [`InvoiceRequest`] that was received.
+		invoice_request: VerifiedInvoiceRequest,
+
+		/// Indicates how the amount is specified across the [`InvoiceRequest`] and the associated [`Offer`] (if any).
+		///
+		/// Use this to determine what amount to pass when calling [`OffersMessageFlow::create_invoice_builder_from_invoice_request`].
+		///
+		/// ### [`InvoiceRequestAmountSource::InvoiceRequestAndOfferAmount`]
+		/// Both the [`InvoiceRequest`] and the [`Offer`] specify amounts.
+		///
+		/// - If the offer uses [`Amount::Currency`], ensure that the request amount reasonably compensates
+		///   based on conversion rates.
+		///
+		/// - If the offer uses [`Amount::Bitcoin`], this implementation ensures the request amount is
+		///   **greater than or equal to** the offer amount before constructing this variant.
+		///
+		/// In either case, use the exact request amount when calling [`OffersMessageFlow::create_invoice_builder_from_invoice_request`].
+		///
+		/// ### [`InvoiceRequestAmountSource::InvoiceRequestOnly`]
+		/// The [`InvoiceRequest`] specifies an amount, and the associated [`Offer`] does not.
+		///
+		/// - You must pass the exact request amount to [`OffersMessageFlow::create_invoice_builder_from_invoice_request`] when responding.
+		///
+		/// ### [`InvoiceRequestAmountSource::OfferOnly`]
+		/// The [`InvoiceRequest`] does not specify an amount. The amount is taken solely from the associated [`Offer`].
+		///
+		/// - If the amount is in [`Amount::Currency`], user must convert it manually into a Bitcoin-denominated amount
+		///   and then pass it to [`OffersMessageFlow::create_invoice_builder_from_invoice_request`].
+		///
+		/// - If the amount is in [`Amount::Bitcoin`], user must provide an amount that is **greater than or equal to**
+		///   the offer amount when calling the builder function.
+		///
+		/// [`Amount::Currency`]: crate::offers::offer::Amount::Currency
+		/// [`Amount::Bitcoin`]: crate::offers::offer::Amount::Bitcoin
+		amount_source: InvoiceRequestAmountSource,
```

Comment:
May not need this here if we can get it directly from `VerifiedInvoiceRequest` as per the other comment.